### ISSUE TITLE
Failed to activate the language-circuitpython package

### ISSUE THREAD

**Issue by RealAbsurdity** [jb3/language-circuitpython#15]
[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.56.0 x64
**Electron**: 9.4.4
**OS**: Microsoft Windows 10 Pro
**Thrown From**: [language-circuitpython](https://github.com/jb3/language-circuitpython) package 0.9.0

### Stack Trace

Failed to activate the language-circuitpython package

```
At The module '\\?\C:\Users\arnie\.atom\packages\language-circuitpython\node_modules\@serialport\bindings\build\Release\bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 73. This version of Node.js requires
NODE_MODULE_VERSION 80. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).

Error: The module '\\?\C:\Users\arnie\.atom\packages\language-circuitpython\node_modules\@serialport\bindings\build\Release\bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 73. This version of Node.js requires
NODE_MODULE_VERSION 80. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
    at process.func [as dlopen] (electron/js2c/asar.js:140:31)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1034:18)
    at Object.func [as .node] (electron/js2c/asar.js:140:31)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module._load (electron/js2c/asar.js:769:28)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149153)
    at bindings (/packages/language-circuitpython/node_modules/bindings/bindings.js:112:48)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/win32.js:1:191)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/win32.js:125:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149837)
    at Object.value [as .js] (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:153385)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module._load (electron/js2c/asar.js:769:28)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149153)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/index.js:6:22)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/index.js:17:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149837)
    at Object.value [as .js] (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:153385)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module._load (electron/js2c/asar.js:769:28)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149153)
    at /packages/language-circuitpython/node_modules/serialport/lib/index.js:2:17)
    at /packages/language-circuitpython/node_modules/serialport/lib/index.js:17:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149837)
```

### Commands

```
     -0:17.3.0 language-circuitpython:toggle-plotter (input.hidden-input)
```

### Non-Core Packages

```
ide-python 1.7.3 
language-circuitpython 0.9.0 
teletype 0.13.4 
```

**Comment by RealAbsurdity** [comment]
Tried the suggested fix in the target directory:

npm install 

new error:
[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.56.0 x64
**Electron**: 9.4.4
**OS**: Microsoft Windows 10 Pro
**Thrown From**: [language-circuitpython](https://github.com/jb3/language-circuitpython) package 0.9.0


### Stack Trace

Failed to activate the language-circuitpython package

```
At The module '\\?\C:\Users\arnie\.atom\packages\language-circuitpython\node_modules\@serialport\bindings\build\Release\bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 83. This version of Node.js requires
NODE_MODULE_VERSION 80. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).

Error: The module '\\?\C:\Users\arnie\.atom\packages\language-circuitpython\node_modules\@serialport\bindings\build\Release\bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 83. This version of Node.js requires
NODE_MODULE_VERSION 80. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
    at process.func [as dlopen] (electron/js2c/asar.js:140:31)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1034:18)
    at Object.func [as .node] (electron/js2c/asar.js:140:31)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module._load (electron/js2c/asar.js:769:28)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149153)
    at bindings (/packages/language-circuitpython/node_modules/bindings/bindings.js:112:48)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/win32.js:1:191)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/win32.js:125:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149837)
    at Object.value [as .js] (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:153385)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module._load (electron/js2c/asar.js:769:28)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149153)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/index.js:6:22)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/index.js:17:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149837)
    at Object.value [as .js] (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:153385)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Module._load (internal/modules/cjs/loader.js:727:14)
    at Function.Module._load (electron/js2c/asar.js:769:28)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149153)
    at /packages/language-circuitpython/node_modules/serialport/lib/index.js:2:17)
    at /packages/language-circuitpython/node_modules/serialport/lib/index.js:17:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.56.0/resources/app/static/<embedded>:11:149837)
```

### Commands

```
  2x -2:06.5.0 language-circuitpython:toggle-serial (input.hidden-input)
```

Non-Core Packages
ide-python 1.7.3 
language-circuitpython 0.9.0 
teletype 0.13.4

**Comment by jb3** [comment]
That's a weird one.

It's possible this was caused by an electron version bump in version 1.56 of Atom, but I've not installed it to reproduce so I can't verify that.

**Comment by RealAbsurdity** [comment]
That was my assumption too and I didn't get a chance to test it yesterday. 

I just downloaded the 1.55 zip version of Atom ran it from an arbitrary location, re-installed language-circuitpython and it worked. So your suggestion led to a workaround for the moment. 

Does that seem to indicate that this is a problem with the included \node_module\serialport  and not with language-circuitpython as a whole?

**Comment by jb3** [comment]
Appreciate you testing that!

I think it's a bit of a mix between us and the serialport library, I'm not sure of the cleanest way to handle native dependencies across multiple versions, or whether there is a clean approach (Native extensions are the cause of most problems with this library).

I'll see if I can get a version bump for serialport this weekend to resolve this on the latest version of atom.

**Comment by RealAbsurdity** [comment]
Thanks for looking into it!

My students I have come to like this plugin to write code in CircuitPyrhon as opposed to Mu because it has all of the Atom features and less serial buffer issues. Really appreciate the plugin and the help with it!

**Comment by jb3** [comment]
Published in version 0.9.1, thanks @dunkmann00 for fixing.

**Comment by dunkmann00** [comment]
@jb3 I think I'm wrong about that serialport update fixing this. When I install this normally it isn't running. Its doing the same thing where the entire Atom window crashes 😞

**Comment by jb3** [comment]
Interesting. It worked for me locally when testing.

Are you getting the same error about the node version being incorrect?

For me locally, running a rebuild with `apm rebuild` seems to fix that, but I don't know how to get client devices to run that on install (or maybe they do?).

APM also seems to be lagging behind NPM functionality, when I run `apm install` it downgrades me to a previous lockfile version, so maybe that ties into it, unsure.

**Comment by jb3** [comment]
Re-opening to diagnose if fixed.

**Comment by dunkmann00** [comment]
For me it is just crashing atom totally...which is a new error I can't say I've seen before. But I do think its related to building/rebuilding.

When I was testing things out on the dev install I have, I got it to work with `electron-rebuild`, but language-circuitpython never needed that so idk what that means. I'm not very knowledgeable about how node builds things so not sure how to tell it what we want it to do.

I'll take another look tomorrow and try `apm-rebuild` and see if that works.

**Comment by dunkmann00** [comment]
@jb3 Just curious, what version of Atom are you running?

**Comment by jb3** [comment]
> For me it is just crashing atom totally...which is a new error I can't say I've seen before. But I do think its related to building/rebuilding.

Interesting, wonder if there are any logs to fetch with that.

> @jb3 Just curious, what version of Atom are you running?

Latest, 1.57.0, I could try the nightly channel.

**Comment by RealAbsurdity** [comment]
I had similar behavior originally with Atom crashing entirely and the only work around was to find the plugin in the local atom install folder and delete it. Without that step after the initial error message it would eventually just crash atom on startup. This is happening again now after I install language-circuitpython and also after reloading if I try to view the installed plugins. With a clean install of Atom 1.57 windows, 64-bit version.

**Comment by dunkmann00** [comment]
I think there might have been a couple of issues going on in this thread, however I think I might have some more info on the more curious case of Atom completely crashing.

For the current version of Atom (1.57), the ABI for modules is 80. This is important for any library that has to compile code, like serialport. The @serialport/bindings library is built for a range of ABIs using [prebuild](https://github.com/prebuild/prebuild). For some reason, the way that the serialport code got compiled for ABI 80, it doesn't work. And I don't think the issue lies in serialport or prebuild, but another library that prebuild is using, [node-abi](https://github.com/lgeiger/node-abi). And from what I can tell, it might not even be an issue with node-abi. The issue might be buried in the electron release that node-abi references.

I'm going to submit a PR to downgrade serialport, and hopefully that will work. It did when I tested it, but we will see.

**Comment by jb3** [comment]
PR merged to supposedly solve this and published in version 0.9.2, so again I'll close the issue, if anyone here notices a regression please feel free to comment and we can re-open or open a new issue.

**Comment by RealAbsurdity** [comment]
It appears this issue has reemerged. Possibly with a new update to Atom... I have a similar error (references this issue) but now with new version numbers:

### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: <https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode>
    * Followed all applicable steps in the debugging guide: <https://flight-manual.atom.io/hacking-atom/sections/debugging/>
    * Checked the FAQs on the message board for common solutions: <https://discuss.atom.io/c/faq>
    * Checked that your issue isn't already filed: <https://github.com/issues?q=is%3Aissue+user%3Aatom>
    * Checked that there is not already an Atom package that provides the described functionality: <https://atom.io/packages>

### Description

Cannot reproduce in safemode as safemode prevents addons from running and the issue is with the language-circuitpython addon...

Language-circuitpython addon cannot start with the error below.

### Steps to Reproduce

1. Open Atom
2. Atom hangs unable to start fully (probably because the CircuitPython Serial was open already
3. Open new Atom window
4. Start CircuitPython Serial
5. Error happens

**Expected behavior:**

Atom should start normally and be able to load the language-CircuitPython Serial addon

**Actual behavior:**

language-CircuitPython serial addon hangs Atom and produces this error

### Versions

**Atom**: 1.63.0 x64
**Electron**: 11.5.0
**OS**: Microsoft Windows 11 Home
**Thrown From**: [language-circuitpython](https://github.com/jb3/language-circuitpython) package 0.9.2


### Stack Trace

Failed to activate the language-circuitpython package

```
At The module '\\?\C:\Users\arnie\.atom\packages\language-circuitpython\node_modules\@serialport\bindings\build\Release\bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 80. This version of Node.js requires
NODE_MODULE_VERSION 85. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).

Error: The module '\\?\C:\Users\arnie\.atom\packages\language-circuitpython\node_modules\@serialport\bindings\build\Release\bindings.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 80. This version of Node.js requires
NODE_MODULE_VERSION 85. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
    at process.func [as dlopen] (electron/js2c/asar_bundle.js:5:1812)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1203:18)
    at Object.func [as .node] (electron/js2c/asar_bundle.js:5:1812)
    at Module.load (internal/modules/cjs/loader.js:992:32)
    at Module._load (internal/modules/cjs/loader.js:885:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12633)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:149229)
    at bindings (/packages/language-circuitpython/node_modules/bindings/bindings.js:112:48)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/win32.js:1:191)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/win32.js:125:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:149913)
    at Object.value [as .js] (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:153507)
    at Module.load (internal/modules/cjs/loader.js:992:32)
    at Module._load (internal/modules/cjs/loader.js:885:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12633)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:149229)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/index.js:6:22)
    at /packages/language-circuitpython/node_modules/@serialport/bindings/lib/index.js:17:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:149913)
    at Object.value [as .js] (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:153507)
    at Module.load (internal/modules/cjs/loader.js:992:32)
    at Module._load (internal/modules/cjs/loader.js:885:14)
    at Function.f._load (electron/js2c/asar_bundle.js:5:12633)
    at Module.require (/app.asar/static/index.js:72:46)
    at require (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:149229)
    at /packages/language-circuitpython/node_modules/serialport/lib/index.js:2:17)
    at /packages/language-circuitpython/node_modules/serialport/lib/index.js:17:3)
    at Module.get_Module._compile (~/AppData/Local/atom/app-1.63.0/resources/app/static/<embedded>:11:149913)
```

### Commands

```
     -3:08.7.0 language-circuitpython:toggle-serial (div.tool-panel.tree-view)
```

### Non-Core Packages

```
language-circuitpython 0.9.2 
```

### Additional Information
